### ISSUE TITLE
Kedge integration

### ISSUE THREAD

**Issue by containscafeine** [fabric8io/fabric8-maven-plugin#1210]
@piyush1594 creating this so this does not slip out of our minds. Can you please update with our discussion from yesterday.

**Comment by piyush1594** [comment]
Right now, we are supporting resource fragments where the user can mention their extra requirements and we override default values by those requirements during the resource generation stage.

Similarly, we can add the integration of Kedge. By doing this, the user can specify Kedge, here not as extra requirements but the overall requirements. Then we can use `Kedge generate` command to generate the kubernetes and openShift resources. If a user specifies a Kedge file, we will replace the resource stage by `kedge generate` command.

Right now, somewhat similar is done using Kompose, a user can specify the docker-compose file and we use Kompose to generate kubernetes yaml.

Comments welcome @containscafeine @hrishin @surajnarwade @tkral @surajssd 

@hrishin Please correct if I am wrong in the concept.

**Comment by containscafeine** [comment]
@piyush1594 I agree with the overall idea :+1: how do we prioritize this? This is a huge bump in user experience with not a lot effort because we simply have to shell out to a binary. ping @pradeepto

**Comment by rohanKanojia** [comment]
@hrishin @nicolaferraro @rhuss: Thoughts????????

**Comment by rhuss** [comment]
Sorry, haven't looked yet at Kedge, but could please someone briefly summarize the difference between Kedge and Kompose (as we already have Kompose support ?)

**Comment by hrishin** [comment]
Kedge tries to simplify writing an application definition using kubernetes native structs. It takes kedgefile as input which has minimalistic  specification of deployment, service etc. and generates a full blown yaml resources as a output. (kind of defining partial resource fagments as we do in FMP but its more simplified) 

http://kedgeproject.org

> could please someone briefly summarize the difference between Kedge and Kompose (as we already have Kompose support ?)

Kompose translates docker compose file to kubernetes resources file. While Kedge tries to simplify writing kubernetes resource yaml. 

@surajnarwade @surajssd @piyush1594 feel please add/correct if something is missing here.

**Comment by surajssd** [comment]
So kedge has it's own way of defining kubernetes app, which is a re-structured version of what you will define in things like Deployment, Service, etc.

Here is an example of one:
http://kedgeproject.org/file-reference/

```yaml
name: httpd

deployments:
- containers:
  - image: centos/httpd

services:
- name: httpd
  type: LoadBalancer
  portMappings: 
    - 8080:80
```

which gets converted to 

```console
$ kedge apply -f httpd.yaml
deployment "httpd" created
service "httpd" created
```

**Comment by containscafeine** [comment]
Worth checking if the project is still actively maintained :)

**Comment by rhuss** [comment]
@surajssd but isn't [kompose](https://github.com/kubernetes/kompose) similar in that respect as it uses plain docker-compose.yml files and creates the appropriate k8s resources ?

So that you can use the simplified compose syntax to create full blow resource descriptors ?

I wonder what the real difference or benefit of kedge over kompose is ...

A kompose file looks like

```yaml
services:

 web:
  image: tuna/docker-counter23
  ports:
    - "5000:5000"
  deploy:
    replicas: 1
    restart_policy:
      condition: any
  labels:
    kompose.service.type: NodePort

 redis:
  image: redis:3.0
  ports:
    - "6379"
```

So it provides maybe an even higher abstraction (abstracting away deployments & service behind one kedge service).

Is there any other difference ?

**Comment by rhuss** [comment]
@cmoulliard sorry I don't understand your answer in the context of this issue about kedge. Could you please elaborate ?

**Comment by lordofthejars** [comment]
What I see is that Kompose is to translate docker compose files to kubernetes resources, meanwhile kedge it is just a way to create kubernetes files in an easy way, without too much hassle. Both projects are different but try to solve the same problem. 

So we can work for this so we also open to other formats FMP and also nice to show in demos so viewers will see that the kedge format simplifies also your experience on having to write less to get k8s resources. 

The only problem I see is how we detect if the yml file is a k8s file, a kompose or kedge. Maybe as an attribute. 

In terms of advantages, well it is another format (developed in RH)  that we support it. I know that there are some projects in F8 that are using it.

**Comment by rhuss** [comment]
> What I see is that Kompose is to translate docker compose files to kubernetes resources, meanwhile kedge it is just a way to create kubernetes files in an easy way, without too much hassle. Both projects are different but try to solve the same problem.

ADAIK, Kompose is a _superset_ to docker compose, adding extra stuff to the syntax, too.

I'm not against Kedge, but could we please confirm that this is still an active project ? From the commit history it seems that the last serious work has been done back in February. Also from the the other metrics (stars, nr. commits), I'm not sure whether we should focus on the current support (kompose, helm ?) and improve that, making it more robust as e.g. the Helm support is still shaky.

The theme for 4.0 is to get out current parts rock solid to make a good basis to explore new land (also thinks like using new ways to build like Jib, Kaniko, maybe KNative support ...). 

tl;dr - I believe Kedge would be a valuable addition, but maybe its a bit early ?

**Comment by rhuss** [comment]
yeah, good idea !

**Comment by rhuss** [comment]
I asked about the status of Kedge here: https://github.com/kedgeproject/kedge/issues/619 as it seems that since @containscafeine and @surajssd have left Red Hat that this project is completely dead.

**Comment by rohanKanojia** [comment]
@rhuss : @surajnarwade @cdrage and @tkral are still in Red Hat. But since the focus has shifted towards odo, kedge doesn't get much bandwidth.

**Comment by rhuss** [comment]
Well, no single commit since six months is for me a clear indicator that it's not active anymore. So let's keep that issue on hold (and should it get stale in 90 days again we might consider to close it)

**Comment by hrishin** [comment]
I think its good to close this issue.

**Comment by rohanKanojia** [comment]
@hrishin : Thanks for your feedback. We've planned to work on odo integration also which makes much more sense, @lordofthejars @dev-gaur , are we good to close this issue now?
